executor/new_distsql.go

```diff
@@ -353,64 +358,45 @@ func (ih *indexWorker) fetchHandles(e *IndexLookUpExecutor, result distsql.Selec
 			case <-finished:
 				return
 			case workCh <- task:
+				e.resultCh <- task
 			}
 		}
 	}
 }
 
-func (ih *indexWorker) close() {
-	ih.wg.Wait()
-}
-
-func (e *IndexLookUpExecutor) waitIndexWorker() {
-	e.indexWorker.close()
+func (worker *indexWorker) close() {
+	worker.wg.Wait()
 }
 
 // tableWorker is used by IndexLookUpExecutor to maintain table lookup background goroutines.
 type tableWorker struct {
-	resultCh chan<- *lookupTableTask
-	wg       sync.WaitGroup
+	wg sync.WaitGroup
 }
 
-// startTableWorker launch some background goroutines which pick tasks from workCh,
-// execute the task and store the results in IndexLookUpExecutor's resultCh.
+// startTableWorker launch some background goroutines which pick tasks from workCh and execute the task.
 func (e *IndexLookUpExecutor) startTableWorker(workCh <-chan *lookupTableTask, finished <-chan struct{}) {
-	resultCh := make(chan *lookupTableTask, atomic.LoadInt32(&LookupTableTaskChannelSize))
-	th := &e.tableWorker
-	th.resultCh = resultCh
-	e.resultChan = resultCh
+	worker := &e.tableWorker
 	lookupConcurrencyLimit := e.ctx.GetSessionVars().IndexLookupConcurrency
-	th.wg.Add(lookupConcurrencyLimit)
+	worker.wg.Add(lookupConcurrencyLimit)
 	for i := 0; i < lookupConcurrencyLimit; i++ {
 		ctx, cancel := goctx.WithCancel(e.ctx.GoCtx())
 		go func() {
-			th.pickAndExecTask(e, workCh, ctx, finished)
+			worker.pickAndExecTask(e, workCh, ctx, finished)
 			cancel()
-			th.wg.Done()
+			worker.wg.Done()
 		}()
 	}
-	go func() {
-		th.wg.Wait()
-		close(th.resultCh)
-	}()
 }
 
-// pickAndExecTask picks tasks from workCh, execute them, and send the result to tableWorker's taskCh.
-func (th *tableWorker) pickAndExecTask(e *IndexLookUpExecutor, workCh <-chan *lookupTableTask, ctx goctx.Context, finished <-chan struct{}) {
+// pickAndExecTask picks tasks from workCh, and execute them.
+func (worker *tableWorker) pickAndExecTask(e *IndexLookUpExecutor, workCh <-chan *lookupTableTask, ctx goctx.Context, finished <-chan struct{}) {
 	for {
 		select {
 		case task, ok := <-workCh:
 			if !ok {
 				return
 			}
-			if task.tasksErr != nil {
-				th.resultCh <- task
-				return
-			}
-
-			// TODO: The results can be simplified when new_distsql.go replace distsql.go totally.
 			e.executeTask(task, ctx)
-			th.resultCh <- task
 		case <-ctx.Done():
```

Comment:
If ctx is done, and then there can be remaining tasks in the `workCh`, which never executes.
Then if the `IndexLookupExecutor` is waiting for the task `done` channel, it will be blocked for ever.

One solution is to remove `ctx.Done()` solution here, waiting for `indexWorker` to close `workCh`.

The `finished` can be removed too, makes the behavior more predictable.
So we can make sure every tasks sent to the `workCh` will be done.

Reply:
It doesn't matter that `workCh` is not consumed.
`indexWorker` would not block by send to that channel, if `ctx.Done()` is true, `indexWorker` will exit and `workCh` would be GCed eventually.

Reply:
But the session thread blocking in `Next`.